drivers/wifi/nrfwifi/src/net_if.c

```diff
@@ -0,0 +1,1179 @@
+	struct nrf_wifi_vif_ctx_zep *vif_ctx_zep = NULL;
+
+	if (!os_vif_ctx) {
+		LOG_ERR("%s: Invalid parameters",
+			__func__);
+		goto out;
+	}
+
+	vif_ctx_zep = os_vif_ctx;
+
+	vif_ctx_zep->if_carr_state = carr_state;
+
+	LOG_DBG("%s: Carrier state: %d", __func__, carr_state);
+
+	k_work_submit(&vif_ctx_zep->nrf_wifi_net_iface_work);
+
+	status = NRF_WIFI_STATUS_SUCCESS;
+
+out:
+	return status;
+}
+
+static bool is_eapol(struct net_pkt *pkt)
+{
+	struct net_eth_hdr *hdr;
+	uint16_t ethertype;
+
+	hdr = NET_ETH_HDR(pkt);
+	ethertype = ntohs(hdr->type);
+
+	return ethertype == NET_ETH_PTYPE_EAPOL;
+}
+#endif /* CONFIG_NRF70_DATA_TX */
+
+enum ethernet_hw_caps nrf_wifi_if_caps_get(const struct device *dev)
+{
+	enum ethernet_hw_caps caps = (ETHERNET_LINK_10BASE_T |
+			ETHERNET_LINK_100BASE_T | ETHERNET_LINK_1000BASE_T);
+
+#ifdef CONFIG_NRF70_TCP_IP_CHECKSUM_OFFLOAD
+	caps |= ETHERNET_HW_TX_CHKSUM_OFFLOAD |
+		ETHERNET_HW_RX_CHKSUM_OFFLOAD;
+#endif /* CONFIG_NRF70_TCP_IP_CHECKSUM_OFFLOAD */
+
+#ifdef CONFIG_NRF70_RAW_DATA_TX
+	caps |= ETHERNET_TXINJECTION_MODE;
+#endif
+#ifdef CONFIG_NRF70_PROMISC_DATA_RX
+	caps |= ETHERNET_PROMISC_MODE;
+#endif
+	return caps;
+}
+
+int nrf_wifi_if_send(const struct device *dev,
+		     struct net_pkt *pkt)
+{
+	int ret = -1;
+#ifdef CONFIG_NRF70_DATA_TX
+	struct nrf_wifi_vif_ctx_zep *vif_ctx_zep = NULL;
+	struct nrf_wifi_ctx_zep *rpu_ctx_zep = NULL;
+
+	if (!dev || !pkt) {
+		LOG_ERR("%s: vif_ctx_zep is NULL", __func__);
+		goto out;
+	}
+
+	vif_ctx_zep = dev->data;
+
+	if (!vif_ctx_zep) {
+		LOG_ERR("%s: vif_ctx_zep is NULL", __func__);
+		goto out;
+	}
+
+	ret = k_mutex_lock(&vif_ctx_zep->vif_lock, K_FOREVER);
+	if (ret != 0) {
+		LOG_ERR("%s: Failed to lock vif_lock", __func__);
+		goto out;
+	}
+
+	rpu_ctx_zep = vif_ctx_zep->rpu_ctx_zep;
+	if (!rpu_ctx_zep || !rpu_ctx_zep->rpu_ctx) {
+		goto unlock;
+	}
+
+#ifdef CONFIG_NRF70_RAW_DATA_TX
+	if ((*(unsigned int *)pkt->frags->data) == NRF_WIFI_MAGIC_NUM_RAWTX) {
+		if (vif_ctx_zep->if_carr_state != NRF_WIFI_FMAC_IF_CARR_STATE_ON) {
+			goto unlock;
+		}
+
+		ret = nrf_wifi_fmac_start_rawpkt_xmit(rpu_ctx_zep->rpu_ctx,
+						      vif_ctx_zep->vif_idx,
+						      net_pkt_to_nbuf(pkt));
+	} else {
+#endif /* CONFIG_NRF70_RAW_DATA_TX */
+		if ((vif_ctx_zep->if_carr_state != NRF_WIFI_FMAC_IF_CARR_STATE_ON) ||
+		    (!vif_ctx_zep->authorized && !is_eapol(pkt))) {
+			goto unlock;
+		}
+
+		ret = nrf_wifi_fmac_start_xmit(rpu_ctx_zep->rpu_ctx,
+					       vif_ctx_zep->vif_idx,
+					       net_pkt_to_nbuf(pkt));
+#ifdef CONFIG_NRF70_RAW_DATA_TX
+	}
+#endif /* CONFIG_NRF70_RAW_DATA_TX */
+unlock:
+	k_mutex_unlock(&vif_ctx_zep->vif_lock);
+#else
+	ARG_UNUSED(dev);
+	ARG_UNUSED(pkt);
+	goto out;
+#endif /* CONFIG_NRF70_DATA_TX */
+
+out:
+	return ret;
+}
+
+#ifdef CONFIG_NRF70_STA_MODE
+static void ip_maddr_event_handler(struct net_mgmt_event_callback *cb,
+				   uint32_t mgmt_event,
+				   struct net_if *iface)
+{
+	struct nrf_wifi_vif_ctx_zep *vif_ctx_zep = NULL;
```

Comment:
Why not use
```
void net_if_mcast_mon_register(struct net_if_mcast_monitor *mon,
                               struct net_if *iface,
                               net_if_mcast_callback_t cb);
```
API here as that was meant to be used from device drivers to get the multicast address register/unregister information.
The problem with using `net_event` API is that the events can disappear if too many events are triggered at the same time.